mtmd: server: Support multimodal data prompt in /completions and /embeddings endpoint of server

65a opened the pull request:
Editing first comment to match current state:

This pull adds support for a multimodal data in the `/completions` (and in a similar fashion, `/embeddings`) API endpoint. Instead of a string, list of tokens, or a mixed string token list as currently supported by that endpoint, this pull add support for a JSON object containing both a `prompt_string` and and a `multimodal_data` field. The client should check the result of `/models` or `/v1/models` for the `multimodal` capability, sending multimodal data to a non-multimodal model will result in a request error.

A singular request example is like this:
```
 {
    ...
    "prompt": { "prompt_string": "string", "multimodal_data": ["base64"] }
    ...
{
```

A multiple (and mixed-type) request would look like:
```
 {
    ...
    "prompt": [
        { "prompt_string": "string", "multimodal_data": ["base64"] },
        "What's up?",
        [ 0, 1, 2, 3, "breakfast" ]
    ]
    ...
{
```
All existing tests pass, new tests are added to cover both the prompt splitting and visual inference. If a prompt is added and the model does not support MTMD, an error is returned. Checking for the `multimodal` in capability `/models` API serves two purposes: it means the model supports it and the llama.cpp is new enough to support this interface. The multimodal part should be a base64-encoded media data supported by libmtmd.

With this approach, other server endpoints can be multimodal in the future relatively easily (Rerank and infill are close but would need additional work and testing, etc). Feedback welcome!

~~Implement a basic way to include multimodal data in the completions endpoint. For now, this just supports directly included data, in base64-encoded format provided as an array of strings under the json key `multimodal_data`. Documentation updated to match. Local testing shows no regression for without-media case, and successful image processing with media provided from a custom client.~~

~~Similar to #14016 but avoids importing the URL fetching logic at all. It could be added later when factored out of the OpenAI-emulation code, but this is simpler for now and avoids need for URL parsing and remote fetch capabilities.~~

Original referenced issue was #13872 

@ngxson ptal, hopefully this works for you.

65a committed:
server : Support multimodal completion and embeddings prompts in JSON format

- Use server_tokens in more places in server and util.cpp
- Convert most functions that used llama_tokens to server_tokens
- Modify input tokenizer to handle JSON objects as subprompts
- Break out MTMD prompt parsing into utility function
- Support JSON objects with multimodal_data arrays for MTMD prompts along with other existing types
- Add capability to model endpoint to indicate if client can send multimodal data
- Add tests.

65a commented on tools/server/README.md:
```diff
@@ -407,6 +408,8 @@ Multiple prompts are also supported. In this case, the completion result will be
   - Strings and sequences of tokens: `["string1", [12, 34, 56]]`
   - Mixed types: `[[12, 34, "string", 56, 78], [12, 34, 56], "string"]`
 
+`multimodal_data`: Optional. Should be an array of strings, containing base64 encoded multimodal data (e.g. images, audio). There must be an identical number of MTMD media markers in the raw prompt which act as placeholders for the data provided to this parameter. The multimodal data files will be substituted in order.
```
It is _currently_ <__media__>, and was <__image__>, but it appears the correct thing to do is use the marker returned by `mtmd_default_marker()`. I assume this means the marker could continue to change.

ngxson commented:
The proposal looks ok but there will be some edge cases:
- What if user enter a list of tokens or mixed list of tokens instead of a raw text prompt?
- What if user enter a list of multiple prompts?

I think proper test cases are required for this PR, similar to `/chat/completion`

65a commented:
I have an idea that might be usable, namely that prompt can now contain an array of JSON objects, like  `"prompt": [{ prompt: "foo", multimodal_data: ["<base64>"] }, { prompt: "bar", multimodal_data: ["<base64>"] }]`. This would be a specific documented option for the prompt field in /completions, and only allow string multimodal prompts (for now). It's a bit verbose and requires more code refactoring, but as a user it makes sense to pair my prompts with their respective data. This would extend the currently supported prompt field options, and not change the top level data JSON object. Also easy to add additional tests for. I'll try it out locally, but does that idea work for you @ngxson ?

65a commented:
Rough draft for the idea here (it compiles and passes existing tests): https://pastebin.com/8zek7ium

Not complete or properly indented, but the idea is to use server_tokens in more places, so that the input tokenizer can branch and use MTMD tokenization where it makes sense to do so. As a side effect, probably got multimodal support in embeddings. Infill needs more work, and rerank would work if I can get the push_back(server_tokens) for server_tokens to work properly, I think.

There are probably better ways to do some of this than I did, feedback welcome.

65a commented:
Improved version of the rough draft that actually works, ignore indentation: https://pastebin.com/R6NdKQPP

This works locally for my use case, and I've started adding tests. There are a few TODOs to make doc ranking and embeddings support multimodal usecases, and I think the oai case can also be streamlined.

The general approach is as described previously: use server_tokens in more places, break out mtmd prompt parsing into a function, and change various input tokenization calls to handle server_tokens instead of llama_tokens.

The request format for multiple prompts would be like this:
```
{
    ...
    "prompt": [
            "Prompt 1",
            [ 1, 2, 3 , 4],
            { "prompt": "What is a tomato?", "multimodal_data": "<base64>"
    ],
    ....
}
```
The JSON entry only supports what `mtmd_tokenize` supports if multimodal_data is present. A single JSON object can also be provided instead of the array of prompts as above. A JSON object could also be provided containing only the prompt in either location, which is handled similarly to normal prompt strings.

65a commented:
Cleaned up the code quite a bit, and fixed the TODO around server_tokens.push_back(server_tokens). Now the tokenize_inputs handling reads a lot cleaner, which is nice.

oobabooga commented:
I have tested this PR and it worked perfectly ✅

Here is a simple test with `google_gemma-3-4b-it-Q4_K_S.gguf`:

<img width="600" alt="print" src="https://github.com/user-attachments/assets/df7dc61d-4507-4ce1-a83b-f51e4422f11e" />

The prompt was:

```
<bos><start_of_turn>user
<__media__>

What do you see here?<end_of_turn>
<start_of_turn>model
```

The details of my UI integration are here https://github.com/oobabooga/text-generation-webui/pull/7027

pnb commented on tools/server/tests/unit/test_vision_completion.py:
```diff
@@ -0,0 +1,48 @@
+        ("What is this: <__media__>\n", IMG_BASE64_0,           True, "(cat)+"), # exceptional, so that we don't cog up the log
+        ("What is this: <__media__>\n", IMG_BASE64_1,           True, "(frog)+"),
+        ("What is this: <__media__>\n", "malformed",              False, None),
+        ("What is this:\n", "base64",        False, None), # non-image data
```
Not that it truly matters but I don't see much consistency in the whitespace for these few lines above, whereas the rest of the code seems more intentional.

65a replied:
Fixed.

ngxson commented on tools/server/tests/unit/test_vision_completion.py:
Can we merge this file to `test_vision_api.py`? We don't have many tests atm, so we should reduce number of files

ngxson commented on tools/server/utils.hpp:
```diff
@@ -1356,3 +1316,132 @@ static std::string fnv_hash(const uint8_t * data, size_t len) {
+    if (json_prompt.is_array() && !json_is_array_with_tokens(json_prompt)) {
+        result.reserve(json_prompt.size());
+        for (const auto & p : json_prompt) {
+            result.push_back(tokenize_input_subprompt(vocab,mctx, p,add_special, parse_special));
```
```suggestion
            result.push_back(tokenize_input_subprompt(vocab,mctx, p, add_special, parse_special));
```

ngxson commented on tools/server/utils.hpp:
```diff
@@ -1356,3 +1316,132 @@ static std::string fnv_hash(const uint8_t * data, size_t len) {
+        // array of tokens
+        llama_tokens tmp = json_prompt.get<llama_tokens>();
+        return server_tokens(tmp, false);
+    } else if (json_prompt.find("prompt") != json_prompt.end()) {
```
```suggestion
    } else if (json_prompt.contains("prompt")) {
```

65a replied:
Done.

ngxson commented on tools/server/utils.hpp:
```diff
@@ -1356,3 +1316,132 @@ static std::string fnv_hash(const uint8_t * data, size_t len) {
+        return server_tokens(tmp, false);
+    } else if (json_prompt.find("prompt") != json_prompt.end()) {
+        // JSON object with prompt key.
+        if (has_mtmd && json_prompt.find("multimodal_data") != json_prompt.end()) {
```
```suggestion
        if (has_mtmd && json_prompt.contains("multimodal_data")) {
```

Or even better, like this:

```
if (json_prompt.contains("multimodal_data")) {
  if (has_mtmd) { ... do the thing .... }
  else throw std::runtime_error("multimodal is not supported by this server");
}
```

ngxson commented on tools/server/server.cpp:
```diff
@@ -4750,22 +4709,22 @@ int main(int argc, char ** argv) {
         }
 
-        llama_tokens tokenized_query = tokenize_input_prompts(ctx_server.vocab, query, /* add_special */ false, true)[0];
+        server_tokens tokenized_query = std::move(tokenize_input_prompts(ctx_server.vocab, ctx_server.mctx, query, /* add_special */ false, true)[0]);
```
I think `std::move` is unnecessary here, the compiler should be good enough to optimize this

ngxson reviewed (commented):
Looking good, can be merge after my comments are all resolved

65a commented:
A few notable updates:
1. Refactored `prompt` and `multimodal_data` to `prompt_string` and `multimodal_data`. This avoids having duplicate keys and opens up the possibility of having things like `prompt_tokens` or `multimodal_urls` in the future.
2. Sending multimodal data to a model without multimodal capability now results in a request error. To avoid that, clients should check `/v1/models` or `/models` for the `multimodal` capabability. Tests added for this new capability as well.
3. Embeddings work and are now documented and tested as multimodal
4. Removed some text about `image_data` in the docs that didn't appear to be implemented yet, more docs for multimodal
5. Made test names better and merged them into the one file.

@ngxson ready for re-review, I won't resolve your comments in case you want to discuss any of the changes.

65a commented on tools/server/utils.hpp:
```diff
@@ -28,6 +28,8 @@
 
 #define DEFAULT_OAICOMPAT_MODEL "gpt-3.5-turbo"
+#define JSON_STRING_PROMPT_KEY "prompt_string"
+#define JSON_MTMD_DATA_KEY "multimodal_data"
```
Done

65a commented on tools/server/utils.hpp:
```diff
@@ -1356,3 +1318,135 @@ static std::string fnv_hash(const uint8_t * data, size_t len) {
+
+            // JSON object with prompt and multimodal key.
+            std::vector<raw_buffer> files;
+            for (const auto& entry : json_prompt.at(JSON_MTMD_DATA_KEY)) {
```
Done.

65a commented on tools/server/utils.hpp:
```diff
@@ -1356,3 +1318,135 @@ static std::string fnv_hash(const uint8_t * data, size_t len) {
+    if (tokenized != 0) {
+        throw std::runtime_error("Failed to tokenize prompt");
+    }
+    auto result = server_tokens(chunks,true);
```
Done.

oobabooga commented:
Thanks for the heads up @65a, I have updated the request! https://github.com/oobabooga/text-generation-webui/commit/e6447cd24acbde845dbb4aa27acfd4c17b5c849c

65a commented:
Looks like the UNDEFINED sanitizer is broken at HEAD by an unrelated change somewhere in ggml :cry:

65a commented:
Can't reproduce sanitizer failure locally, hopefully pushing an updated commit message can retrigger CI.

ngxson commented on tools/server/tests/unit/test_vision_api.py:
```diff
@@ -10,21 +10,48 @@
 
+def test_models_supports_multimodal_capability():
+    global server
+    server.start(timeout_seconds=60) # vision model may take longer to load due to download size
```
```suggestion
    server.start()
```

the model is not big, just 50MB. it should be a problem on your side

ngxson commented on tools/server/tests/unit/test_vision_api.py:
```diff
@@ -10,21 +10,48 @@
+
+def test_v1_models_supports_multimodal_capability():
+    global server
+    server.start(timeout_seconds=60) # vision model may take longer to load due to download size
```
```suggestion
    server.start()
```

65a replied:
Done.

ngxson commented on tools/server/tests/unit/test_vision_api.py:
```diff
@@ -58,3 +85,62 @@ def test_vision_chat_completion(prompt, image_url, success, re_content):
+    global server
+    server.server_embeddings=True
+    server.n_batch=512
+    server.start(timeout_seconds=60) # vision model may take longer to load due to download size
```
```suggestion
    server.start()
```

ngxson commented on tools/server/utils.hpp:
```diff
@@ -123,6 +123,19 @@ static bool json_is_array_of_mixed_numbers_strings(const json & data) {
 }
 
+// does array have any individual integers/tokens?
+static bool json_is_array_with_tokens(const json & data) {
```
```suggestion
static bool json_is_array_contains_number(const json & data) {
```

65a replied:
Done.

65a commented on tools/server/tests/unit/test_vision_api.py:
```diff
@@ -58,3 +85,62 @@ def test_vision_chat_completion(prompt, image_url, success, re_content):
+)
+def test_vision_completion(prompt, image_data, success, re_content):
+    global server
+    server.start(timeout_seconds=60) # vision model may take longer to load due to download size
```
Done.

65a commented:
Looks good now?
